esp_commands/README.md

```diff
@@ -0,0 +1,183 @@
+# ESP Commands
+
+The `esp_commands` component provides a flexible command registration and execution framework for ESP-IDF applications.  
+It allows applications to define console-like commands with metadata (help text, hints, glossary entries) and register them dynamically or statically.
+
+---
+
+## Features
+
+- Define commands with:
+  - Command name
+  - Group categorization
+  - Help text
+  - Optional hints and glossary callbacks
+- Register commands at runtime or at compile-time (via section placement macros).
+- Execute commands from command line strings.
+- Provide command completion, hints, and glossary callback registration mechanism.
+- Create and manage subsets of commands (command sets).
+- Customizable configuration for command parsing and hint display.
+
+---
+
+## Configuration
+
+The component is initialized with a configuration struct:
+
+```c
+esp_commands_config_t config = ESP_COMMANDS_CONFIG_DEFAULT();
+esp_commands_update_config(&config);
+```
+
+- `write_func`: The custom write function used by esp_commands to output data (default to posix write is not specified)
+- `max_cmdline_length`: Maximum command line buffer length (bytes).
+- `max_cmdline_args`: Maximum number of arguments parsed.
+- `hint_color`: ANSI color code used for hints.
+- `hint_bold`: Whether hints are displayed in bold.
+
+---
+
+## Defining Commands
+
+### Command Structure
+
+A command is described by the `esp_command_t` struct:
+
+```c
+typedef struct esp_command {
+    const char *name;                     /*!< Command name */
+    const char *group;                    /*!< Group/category */
+    const char *help;                     /*!< Short help text */
+    esp_command_func_t func;              /*!< Command implementation */
+    void *func_ctx;                       /*!< User context */
+    esp_command_hint_t hint_cb;           /*!< Hint callback */
+    esp_command_glossary_t glossary_cb;   /*!< Glossary callback */
+} esp_command_t;
+```
+
+### Static Registration
+
+Use the `ESP_COMMAND_REGISTER` macro to register a command at compile time:
+
+```c
+static int my_cmd(void *context, esp_commands_exec_arg_t *cmd_arg, int argc, char **argv) {
+    printf("Hello from my_cmd!\n");
+    return 0;
+}
+
+ESP_COMMAND_REGISTER(my_cmd, tools, "Prints hello", my_cmd, NULL, NULL, NULL);
+```
+
+This places the command into the `.esp_commands` section in flash.
+
+### Dynamic Registration
+
+Commands can also be registered/unregistered at runtime:
+
+```c
+esp_command_t cmd = {
+    .name = "echo",
+    .group = "utils",
+    .help = "Echoes arguments back",
+    .func = echo_func,
```

Comment:
Do we use a different function from my_cmd for a reason here? Seems clearer if both registration examples use the same one.